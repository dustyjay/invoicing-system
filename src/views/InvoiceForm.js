import React, { useEffect, useState } from 'react';
import FormInput from '../shared/Input';
import { formatInvoiceNumber, userInfo } from '../shared/invoices';
import { Tab, Tabs, TabList, TabPanel } from 'react-tabs';
import 'react-tabs/style/react-tabs.css';

const InvoiceForm = ({ invoice, index, onEdit }) => {
  const [formData, setFormData] = useState({});

  useEffect(() => {
    setFormData(invoice);
  }, [invoice]);

  const setFormValues = (name, value) => {
    setFormData({ ...formData, [name]: value });
  };

  const saveChanges = () => {
    onEdit(formData);
  };

  const setUserValues = (name, value) => {
    const newUserData = { ...formData.userData, [name]: value };
    setFormValues('userData', newUserData);
  };

  return (
    <React.Fragment>
      <form className="invoice-form">
        <Tabs>
          <TabList>
            <Tab>Client Details</Tab>
            <Tab>Personal Details</Tab>
            <Tab>Account Details</Tab>
            <Tab>Other</Tab>
          </TabList>

          <TabPanel>
            <div className="form-section__content">
              <FormInput
                value={formData.clientName}
                placeholder="Enter client name here"
                name="clientname"
                label="Client Name"
                onInput={e => setFormValues('clientName', e)}
                onBlur={saveChanges}
              />
              <FormInput
                value={formData.dateIssued}
                placeholder="Date issued"
                type="date"
                name="dateissued"
                label="Date Issued"
                readOnly
              />
              <FormInput
                value={formatInvoiceNumber(index)}
                placeholder="Enter invoice number"
                name="invoicenumber"
                label="Invoice No. #"
                readOnly
              />
            </div>
          </TabPanel>
          <TabPanel>
            {formData.userData && (
              <div className="form-section__content">
                <FormInput
                  value={formData.userData.name}
                  placeholder="Enter your name here"
                  name="yourname"
                  label="Your Name"
                  onInput={e => setUserValues('name', e)}
                  onBlur={saveChanges}
                />
                <FormInput
                  value={formData.userData.email}
                  placeholder="Enter your email here"
                  name="youremail"
                  label="Email Address"
                  onInput={e => setUserValues('email', e)}
                  onBlur={saveChanges}
                />
                <FormInput
                  value={formData.userData.phone}
                  placeholder="Enter your phone number here"
                  type="tel"
                  name="yourphone"
                  label="Phone Number"
                  onInput={e => setUserValues('phone', e)}
                  onBlur={saveChanges}
                />
                <FormInput
                  value={formData.userData.website}
                  placeholder="Enter your website here"
                  type="url"
                  name="yourwebsite"
                  label="Website"
                  onInput={e => setUserValues('website', e)}
                  onBlur={saveChanges}
                />
                <FormInput
                  value={formData.userData.address}
                  placeholder="Enter your home address here"
                  name="youraddress"
                  label="Home Address"
                  onInput={e => setUserValues('address', e)}
                  onBlur={saveChanges}
                />
                <FormInput
                  value={formData.userData.town}
                  placeholder="Enter your town here"
                  name="yourtown"
                  label="Town"
                  onInput={e => setUserValues('town', e)}
                  onBlur={saveChanges}
                />
                <FormInput
                  value={formData.userData.city}
                  placeholder="Enter your city here"
                  name="yourcity"
                  label="City"
                  onInput={e => setUserValues('city', e)}
                  onBlur={saveChanges}
                />
                <FormInput
                  value={formData.userData.postalCode}
                  placeholder="Enter your postalCode here"
                  type="number"
                  name="yourpostalcode"
                  label="Postal Code"
                  onInput={e => setUserValues('postalCode', e)}
                  onBlur={saveChanges}
                />
              </div>
            )}
          </TabPanel>
          <TabPanel>
            {formData.userData && (
              <div className="form-section__content">
                <FormInput
                  value={formData.userData.accountNumber}
                  placeholder="Enter your account number here"
                  type="number"
                  name="accountnumber"
                  label="Account Number"
                  onInput={e => setUserValues('accountNumber', e)}
                  onBlur={saveChanges}
                />
                <FormInput
                  value={formData.userData.sortCode}
                  placeholder="Enter your bank sort code here"
                  type="number"
                  name="sortcode"
                  label="Sort code"
                  onInput={e => setUserValues('sortCode', e)}
                  onBlur={saveChanges}
                />
              </div>
            )}
          </TabPanel>
          <TabPanel>
            <div className="form-section__content">
              <FormInput
                value={formData.discount}
                placeholder="Enter discount percentage"
                name="discount"
                label="Discount (%)"
                onInput={e => setFormValues('discount', e)}
                onBlur={saveChanges}
              />
              <FormInput
                value={formData.tax}
                placeholder="Enter tax percentage"
                name="tax"
                label="Tax Rate (%)"
                onInput={e => setFormValues('tax', e)}
                onBlur={saveChanges}
              />
              <FormInput
                value={formData.dueDate}
                placeholder="Select due date"
                type="date"
                name="duedate"
                label="Due date"
                onInput={e => setFormValues('dueDate', e)}
                onBlur={saveChanges}
              />
            </div>
          </TabPanel>
        </Tabs>
      </form>
    </React.Fragment>
  );
};

export default InvoiceForm;
