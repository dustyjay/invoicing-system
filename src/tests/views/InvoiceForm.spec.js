import React from 'react';
import { render, cleanup } from '@testing-library/react';
import InvoiceForm from '../../views/InvoiceForm';
import { INVOICES } from '../testData';

const customRender = (ui, options) =>
  render(ui, {
    wrapper: props => <InvoiceForm {...props} {...options.wrapperProps} />,
    ...options
  });

const INVOICE_INDEX = 0;

describe('Invoice Form', () => {
  let wrapper;
  let container;

  beforeEach(() => {
    wrapper = customRender(InvoiceForm, {
      wrapperProps: {
        invoice: INVOICES[INVOICE_INDEX],
        index: INVOICE_INDEX,
        onEdit: false
      }
    });
    container = wrapper.container;
  });

  afterEach(() => {
    cleanup();
  });

  test('Invoice Form matches snapshot', () => {
    const tree = wrapper.container;
    expect(tree).toMatchSnapshot();
  });

  test('has the right invoice values', () => {
    const invoice = INVOICES[INVOICE_INDEX];
    const clientNameInput = container.querySelector('#clientname');
    expect(clientNameInput).toHaveValue(invoice.clientName);
  });
});
