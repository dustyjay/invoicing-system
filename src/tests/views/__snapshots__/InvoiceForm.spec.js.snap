// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Invoice Form Invoice Form matches snapshot 1`] = `
<div>
  <form
    class="invoice-form"
  >
    <div
      class="react-tabs"
      data-tabs="true"
    >
      <ul
        class="react-tabs__tab-list"
        role="tablist"
      >
        <li
          aria-controls="react-tabs-1"
          aria-disabled="false"
          aria-selected="true"
          class="react-tabs__tab react-tabs__tab--selected"
          id="react-tabs-0"
          role="tab"
          tabindex="0"
        >
          Client Details
        </li>
        <li
          aria-controls="react-tabs-3"
          aria-disabled="false"
          aria-selected="false"
          class="react-tabs__tab"
          id="react-tabs-2"
          role="tab"
        >
          Personal Details
        </li>
        <li
          aria-controls="react-tabs-5"
          aria-disabled="false"
          aria-selected="false"
          class="react-tabs__tab"
          id="react-tabs-4"
          role="tab"
        >
          Account Details
        </li>
        <li
          aria-controls="react-tabs-7"
          aria-disabled="false"
          aria-selected="false"
          class="react-tabs__tab"
          id="react-tabs-6"
          role="tab"
        >
          Other
        </li>
      </ul>
      <div
        aria-labelledby="react-tabs-0"
        class="react-tabs__tab-panel react-tabs__tab-panel--selected"
        id="react-tabs-1"
        role="tabpanel"
      >
        <div
          class="form-section__content"
        >
          <div
            class="input__box"
          >
            <label
              class="input__label"
              for="clientname"
            >
              Client Name
            </label>
            <input
              class="input"
              id="clientname"
              placeholder="Enter client name here"
              type="text"
              value="Jurassic Park"
            />
          </div>
          <div
            class="input__box"
          >
            <label
              class="input__label"
              for="dateissued"
            >
              Date Issued
            </label>
            <input
              class="input"
              disabled=""
              id="dateissued"
              placeholder="Date issued"
              readonly=""
              type="date"
              value="24 May, 2031"
            />
          </div>
          <div
            class="input__box"
          >
            <label
              class="input__label"
              for="invoicenumber"
            >
              Invoice No. #
            </label>
            <input
              class="input"
              disabled=""
              id="invoicenumber"
              placeholder="Enter invoice number"
              readonly=""
              type="text"
              value="00000"
            />
          </div>
        </div>
      </div>
      <div
        aria-labelledby="react-tabs-2"
        class="react-tabs__tab-panel"
        id="react-tabs-3"
        role="tabpanel"
      />
      <div
        aria-labelledby="react-tabs-4"
        class="react-tabs__tab-panel"
        id="react-tabs-5"
        role="tabpanel"
      />
      <div
        aria-labelledby="react-tabs-6"
        class="react-tabs__tab-panel"
        id="react-tabs-7"
        role="tabpanel"
      />
    </div>
  </form>
</div>
`;
